src/coreComponents/constitutive/fluid/MultiPhaseMultiComponentFluid.hpp

```diff
@@ -267,31 +298,600 @@ class MultiPhaseMultiComponentFluid : public MultiFluidBase
   {
     static constexpr char const * flashModelParaFileString() { return "flashModelParaFile"; }
     static constexpr char const * phasePVTParaFilesString() { return "phasePVTParaFiles"; }
-  };
-
+  } viewKeysMultiPhaseMultiComponentFluid;
 
 protected:
-  virtual void postProcessInput() override;
 
-  virtual void initializePostSubGroups() override;
+  virtual void postProcessInput() override;
 
 private:
 
   void createPVTModels();
 
-  // phase PVT parameter filenames
+  /// Names of the files defining the viscosity and density models
   path_array m_phasePVTParaFiles;
 
+  /// Name of the file defining the flash model
   Path m_flashModelParaFile;
 
-  // number of entries corrosponds to number of phases
-  std::vector< std::shared_ptr< PVTProps::PVTFunction const > > m_phaseDensityFuns;
-  std::vector< std::shared_ptr< PVTProps::PVTFunction const > > m_phaseViscosityFuns;
+  /// Index of the liquid phase
+  localIndex m_p1Index;
+
+  /// Index of the gas phase
+  localIndex m_p2Index;
+
+  /// Pointer to the brine density model
+  std::unique_ptr< P1DENS > m_p1Density;
+
+  /// Pointer to the brine viscosity model
+  std::unique_ptr< P1VISC > m_p1Viscosity;
+
+  /// Pointer to the CO2 density model
+  std::unique_ptr< P2DENS > m_p2Density;
+
+  /// Pointer to the CO2 viscosity model
+  std::unique_ptr< P2VISC > m_p2Viscosity;
+
+  /// Pointer to the flash model
+  std::unique_ptr< FLASH > m_flash;
+
+  /// Pointer to the brine density model
+  array1d< typename P1DENS::KernelWrapper > m_p1DensityWrapper;
+
+  /// Pointer to the brine viscosity model
+  array1d< typename P1VISC::KernelWrapper > m_p1ViscosityWrapper;
+
+  /// Pointer to the CO2 density model
+  array1d< typename P2DENS::KernelWrapper > m_p2DensityWrapper;
 
-  std::shared_ptr< PVTProps::FlashModel const > m_flashModel;
+  /// Pointer to the CO2 viscosity model
+  array1d< typename P2VISC::KernelWrapper > m_p2ViscosityWrapper;
+
+  /// Pointer to the flash model
+  array1d< typename FLASH::KernelWrapper > m_flashWrapper;
```

Comment:
You should really document that these `array1d` are tricks (and explain the issue once somewhere.)